src/easy3d.c

```diff
@@ -0,0 +1,203 @@
+        DC_FlushRange(*resource, (*resource)->fileSize);
+        NNS_G3dResDefaultSetup(*resource);
+    }
+
+    *outModel = NNS_G3dGetMdlByIdx(NNS_G3dGetMdlSet(*resource), 0);
+}
+
+void Easy3D_InitRenderObjFromPath(const u8 heapID, const char *path, NNSG3dRenderObj *obj, NNSG3dResMdl **outModel, NNSG3dResFileHeader **outResource)
+{
+    Easy3D_LoadModelFromPath(heapID, path, outModel, outResource);
+    NNS_G3dRenderObjInit(obj, *outModel);
+}
+
+void Easy3D_InitRenderObjFromResource(NNSG3dRenderObj *renderObj, NNSG3dResMdl **model, NNSG3dResFileHeader **resource)
+{
+    Easy3D_LoadModelFromResource(model, resource);
+    GF_ASSERT(model);
+    NNS_G3dRenderObjInit(renderObj, *model);
+}
+
+BOOL Easy3D_IsTextureUploadedToVRam(NNSG3dResTex *texture)
+{
+    return (texture->texInfo.flag & NNS_G3D_RESTEX_LOADED) 
+        || (texture->tex4x4Info.flag & NNS_G3D_RESTEX4x4_LOADED);
+}
+
+void Easy3D_DrawRenderObj(NNSG3dRenderObj *renderObj, const VecFx32 *pos, const MtxFx33 *rot, const VecFx32 *scale)
+{
+    NNS_G3dGlbSetBaseTrans(pos);
+    NNS_G3dGlbSetBaseRot(rot);
+    NNS_G3dGlbSetBaseScale(scale);
+    NNS_G3dGlbFlush();
+    NNS_G3dDraw(renderObj);
+}
+
+void Easy3D_DrawRenderObjSimple(NNSG3dRenderObj *renderObj, const VecFx32 *pos, const MtxFx33 *rot, const VecFx32 *scale)
+{
+    NNS_G3dGlbSetBaseTrans(pos);
+    NNS_G3dGlbSetBaseRot(rot);
+    NNS_G3dGlbSetBaseScale(scale);
+    NNS_G3dGlbFlush();
+    NNS_G3dDraw1Mat1Shp(renderObj->resMdl, 0, 0, TRUE);
+    NNS_G3dGeFlushBuffer();
+}
+
+static GenericPointerData *Unk_021BFB0C = NULL;
+
+void Easy3D_Init(const u8 heapID)
+{
+    Unk_021BFB0C = sub_02024220(heapID, 0, 2, 0, 4, Easy3D_EngineSetup);
+}
+
+static void Easy3D_EngineSetup(void)
+{
+    GXLayers_EngineAToggleLayers(GX_PLANEMASK_BG0, TRUE);
+
+    G2_SetBG0Priority(1);
+    G3X_SetShading(GX_SHADING_TOON);
+    G3X_AntiAlias(TRUE);
+    G3X_AlphaTest(FALSE, 0);
+    G3X_AlphaBlend(TRUE);
+    G3X_EdgeMarking(FALSE);
+    G3X_SetFog(FALSE, GX_FOGBLEND_COLOR_ALPHA, GX_FOGSLOPE_0x8000, 0);
+    G3X_SetClearColor(GX_RGB(0, 0, 0), 0, 0x7fff, 63, FALSE);
+    G3_ViewPort(0, 0, 255, 191);
+}
+
+void Easy3D_Shutdown(void)
+{
+    sub_020242C4(Unk_021BFB0C);
+}
+
+BOOL Easy3D_BindTextureToResource(void *resource, NNSG3dResTex *texture)
+{
+    NNSG3dResMdlSet *modelSet = NNS_G3dGetMdlSet((NNSG3dResFileHeader *)resource);
+    NNS_G3D_NULL_ASSERT(modelSet);
+
+    if (texture) {
+        BOOL v1 = NNS_G3dBindMdlSet(modelSet, texture);
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+BOOL Easy3D_UploadTextureToVRam(NNSG3dResTex *texture)
+{
+    if (texture == NULL) {
+        return FALSE;
+    }
+
+    texture->texInfo.vramKey = 0;
+
+    u32 texRequiredSize, 
+        tex4x4RequiredSize, 
+        paletteRequiredSize;
+    BOOL texOk = TRUE;
+    BOOL tex4x4Ok = TRUE;
+    BOOL paletteOk = TRUE;
+    NNSG3dTexKey texKey;
+    NNSG3dTexKey tex4x4Key;
+    NNSG3dPlttKey paletteKey;
+
+    if (texture) {
```

Comment:
**question:** Is this `if` necessary? Does the compiler strip it due to the preceding sentinel?

Reply:
Yeah it doesn't match without it